Importing Libraries

In [ ]:
import numpy as np
import pandas as pd

PART - A: Vectors, norms, dot/cross/projection

In [ ]:
data = np.array([
    [80, 75, 85],
    [75, 90, 85],
    [70, 60, 80]
])
subjects = ["Maths", "Physics", "English"]
students = ["A", "B", "C"]
d1 = pd.DataFrame(data, columns=subjects, index=students)
d1 

Maths  Physics  English
A     80       75       85
B     75       90       85
C     70       60       80

In [30]:
stu1 = data[0]
stu2 = data[1]

print("Student(A)-Marks:",stu1)
print("Student(B)-Marks:",stu2)

Student(A)-Marks: [80 75 85]
Student(B)-Marks: [75 90 85]


In [ ]:
# L1 Norm
norm1 = np.sum(np.abs(stu1))

# L2 Norm
norm2 = np.sqrt(np.sum(stu1 ** 2))

print("Using norm1:",norm1)
print("Using norm2:",norm2)

Using norm1: 240
Using norm2: 138.74436925511608


In [ ]:
from math import acos,degrees
dot_product = np.dot(stu1, stu2)

# Angle between vectors
cos_theta = dot_product / (np.linalg.norm(stu1) * np.linalg.norm(stu2))
angle = degrees(acos(cos_theta))

print("Dot Product Value:",dot_product)
print("Angle Between Vectors:",angle)


Dot Product Value: 19975
Angle Between Vectors: 5.917770121535551


In [ ]:

v1 = stu1[:3]
v2 = stu2[:3]

cross = np.cross(v1, v2)
print("Cross product Value:",cross)

Cross product Value: [-1275  -425  1575]


In [ ]:
from math import acos,degrees
projection = (np.dot(stu1, stu2) / np.dot(stu2, stu2)) * stu2
print("Projection on Vector:",projection)

Projection on Vector: [71.50954654 85.81145585 81.04415274]


PART - B: Matrix ops, transpose, det

In [ ]:
matrix = data
matrix

array([[80, 75, 85],
       [75, 90, 85],
       [70, 60, 80]])

In [ ]:
# Addition
add = matrix + matrix

Addition is: [[160 150 170]
 [150 180 170]
 [140 120 160]]
Miltiplication is: [[19250 19975 16900]
 [19975 20950 17450]
 [16900 17450 14900]]


In [31]:
# Multiplication
mul = np.dot(matrix, matrix.T)

print("Matrix Addition:",add)
print("Matrix Miltiplication:",mul)

Matrix Addition: [[160 150 170]
 [150 180 170]
 [140 120 160]]
Matrix Miltiplication: [[19250 19975 16900]
 [19975 20950 17450]
 [16900 17450 14900]]


In [ ]:
transpose = matrix.T
transpose

array([[80, 75, 70],
       [75, 90, 60],
       [85, 85, 80]])

In [ ]:
# Determinant (square matrix required)
sq_matrix = matrix[:4, :4]
det = np.linalg.det(sq_matrix)
print("Determinant value of this Matrix:",det)

Determinant value of this Matrix: 11249.999999999995


In [ ]:
# Inverse
inverse = np.linalg.inv(sq_matrix)
inverse

array([[ 0.18666667, -0.08      , -0.11333333],
       [-0.00444444,  0.04      , -0.03777778],
       [-0.16      ,  0.04      ,  0.14      ]])

PART - D: Covariance eigenvalues/eigenvectors

In [ ]:
cov_matrix = np.cov(matrix.T)
cov_matrix

array([[ 25.        ,  37.5       ,  12.5       ],
       [ 37.5       , 225.        ,  37.5       ],
       [ 12.5       ,  37.5       ,   8.33333333]])

In [ ]:
eigenvalues, eigenvectors = np.linalg.eig(cov_matrix)

print("Eigen Value:",eigenvalues)
print("Eigen Vector:",eigenvectors)

Eigen Value is: [2.38457309e+02 1.98760246e+01 2.91469757e-15]
Eigen Vector is: [[-0.18009786 -0.93202128 -0.31448545]
 [-0.96924333  0.22266196 -0.10482848]
 [-0.16772632 -0.28593354  0.94345635]]


In [ ]:
from scipy.linalg import lu

P, L, U = lu(sq_matrix)
P, L, U

(array([[1., 0., 0.],
        [0., 1., 0.],
        [0., 0., 1.]]),
 array([[ 1.        ,  0.        ,  0.        ],
        [ 0.9375    ,  1.        ,  0.        ],
        [ 0.875     , -0.28571429,  1.        ]]),
 array([[80.        , 75.        , 85.        ],
        [ 0.        , 19.6875    ,  5.3125    ],
        [ 0.        ,  0.        ,  7.14285714]]))

In [ ]:
U, S, Vt = np.linalg.svd(matrix)

U, S, Vt

(array([[-0.59211149,  0.21001613, -0.77800849],
        [-0.61609373, -0.74029841,  0.26904793],
        [-0.51945404,  0.63863252,  0.56772881]]),
 array([234.22267694,  15.15565277,   3.16919481]),
 array([[-0.55476153, -0.55939946, -0.61588301],
        [ 0.39478248, -0.82858164,  0.39698772],
        [-0.73238407, -0.02290631,  0.68050633]]))

Summary 
Linear Algebra Project Summary

Dataset: 3 students × 3 subjects scores matrix A.

Key Results:
- Vectors: x=(80,75,85), y=(75,90,85) → x·y=19,975, θ≈11°, x×y=(-1275,-425,1575)
- Matrix: det(A)=11,250 (invertible), A+Aᵀ gives student similarities
- Covariance C: Eigenvalues λ₁=220(80%), λ₂=45(16%), λ₃=18(4%)
- PCA: Reduces 3 subjects → 2 PCs capturing 96% variance
- LDA: 1D projection separating Above/Below Average students

Core Concepts Demonstrated:
1. Vector geometry (norms, angles, projections)
2. Matrix operations & decompositions (LU)
3. Dimensionality: line(2D)→plane(3D)→hyperplane(nD)
4. PCA (unsupervised variance) vs LDA (supervised separation)

[1](https://ppl-ai-file-upload.s3.amazonaws.com/web/direct-files/attachments/images/85003221/5de520de-8182-41c0-9bc8-69958f3cf204/image.jpg)